src/librustc_incremental/persist/load.rs

```diff
@@ -0,0 +1,221 @@
+                &format!("could not load dep-graph from `{}`: {}",
+                         path.display(), err));
+            return;
+        }
+    }
+
+    match decode_dep_graph(tcx, Doc::new(&data)) {
+        Ok(dirty) => dirty,
+        Err(err) => {
+            tcx.sess.bug(
+                &format!("decoding error in dep-graph from `{}`: {}",
+                         path.display(), err));
+        }
+    }
+}
+
+pub fn decode_dep_graph<'tcx, 'doc>(tcx: &ty::TyCtxt<'tcx>, doc: rbml::Doc<'doc>)
+                                    -> DecodeResult<()>
+{
+    // First load the directory, which maps the def-ids found
+    // elsewhere into `DefPath`. We can then refresh the `DefPath` to
+    // obtain updated def-ids.
+    let directory = {
+        let directory_doc = reader::get_doc(doc, DIRECTORY_TAG);
+        let mut decoder = Decoder::new(directory_doc);
+        try!(DefIdDirectory::decode(&mut decoder))
+    };
+
+    debug!("decode_dep_graph: directory = {:#?}", directory);
+
+    // Retrace those paths to find their current location (if any).
+    let retraced = directory.retrace(tcx);
+
+    debug!("decode_dep_graph: retraced = {:#?}", retraced);
+
+    // Deserialize the dep-graph (which will include DefPathIndex entries)
+    let serialized_dep_graph = {
+        let dep_graph_doc = reader::get_doc(doc, DEP_GRAPH_TAG);
+        let mut decoder = Decoder::new(dep_graph_doc);
+        try!(SerializedDepGraph::decode(&mut decoder))
+    };
+
+    debug!("decode_dep_graph: serialized_dep_graph = {:#?}", serialized_dep_graph);
+
+    // Compute the set of Hir nodes whose data has changed.
+    let mut dirty_nodes =
+        initial_dirty_nodes(tcx, &serialized_dep_graph.hashes, &retraced);
+
+    debug!("decode_dep_graph: initial dirty_nodes = {:#?}", dirty_nodes);
+
+    // Find all DepNodes reachable from that core set. This loop
+    // iterates repeatedly over the list of edges whose source is not
+    // known to be dirty (`clean_edges`). If it finds an edge whose
+    // source is dirty, it removes it from that list and adds the
+    // target to `dirty_nodes`. It stops when it reaches a fixed
+    // point.
+    let clean_edges = compute_clean_edges(&serialized_dep_graph.edges,
+                                          &retraced,
+                                          &mut dirty_nodes);
+
+    // Add synthetic `foo->foo` edges for each clean node `foo` that
+    // we had before. This is sort of a hack to create clean nodes in
+    // the graph, since the existence of a node is a signal that the
+    // work it represents need not be repeated.
+    let clean_nodes =
+        serialized_dep_graph.nodes
+                            .iter()
+                            .filter_map(|&node| retraced.map(node))
+                            .filter(|node| !dirty_nodes.contains(node))
+                            .map(|node| (node, node));
+
+    // Add nodes and edges that are not dirty into our main graph.
```

Comment:
Since we are mutating an existing dep-graph, it would be good to somewhere document what state we expect this dep-graph to be in.